File: src/clients/hazelcast-client.js
const { HazelcastClient: HClient } = require("hazelcast-client/lib/HazelcastClient");
const fs = require('fs');


class HazelcastClient {

  constructor(numOfInstances = 1, resultFilename = 'results-hazelcast.txt', totalOperations = 10000) {
    this.config = {
      ...this.generateConfig(numOfInstances),
      customLogger: new SilentLogger()
    }
    this.numOfInstances = numOfInstances;
    this.resultFilename = resultFilename;
    this.totalOperations = totalOperations;
  }

  generateConfig(numOfInstances) {
    const config = {
      network: {
        clusterMembers: []
      }
    }
  
    for (let i = 0; i < numOfInstances; i++) {
      config.network.clusterMembers.push(`127.0.0.1:570${i}`);
    }
  
    return config;
  }

  async connect() {
    this.client = await HClient.newHazelcastClient(this.config);
    this.map = await this.client.getMap('benchmark-map');
    console.log("Connected to Hazelcast");
    return this.client;
  }

  async disconnect() {
    await this.client.shutdown();
    console.log("Hazelcast Client Disconnected");
  }

  async cleanup() {
    for (let key of this.keys) {
      await this.map.delete(key);
    }
  }

  benchmarkWrite(dataSize, show = false) {
    return new Promise((resolve) => {
      let data = Buffer.alloc(dataSize, 'a').toString(); 
      let totalOperations = 10000;
      let totalTimeTaken = 0;
      this.keys = [];
      const operations = [];
  
      for (let i = 0; i < totalOperations; i++) {
        const key = `key-${dataSize}-${i}`;
        this.keys.push(key);
        
        operations.push(this.map.set(key, data));  
      }
      const start = process.hrtime.bigint();
      Promise.all(operations).then(() => {
        const end = process.hrtime.bigint();
    
        totalTimeTaken += Number(end - start) / 1e6;

        const averageTime = totalTimeTaken / totalOperations;
        if(show) {
          console.info(`Hazelcast write for size ${dataSize} bytes: ${averageTime.toFixed(6)} ms per operation`);
          fs.appendFile(this.resultFilename, 'Hazelcast, WRITE, ' + this.numOfInstances + ', ' + dataSize + ', ' + averageTime.toFixed(6) + '\n', (err) => {
            if(err) console.error(err);
          });
        }

        resolve();
      });
    })
  }

  benchmarkRead(dataSize, show = false) {
    return new Promise((resolve) => {
      const keys = this.keys
                    .map(val => ({ val, sort: Math.random() }))
                    .sort((a, b) => a.sort - b.sort)
                    .map(obj => obj.val);

      let operations = keys.map(key => this.map.get(key));
      
      const start = process.hrtime.bigint();
      Promise.all(operations).then(() => {
        const end = process.hrtime.bigint();
    
        const averageTime = (Number(end - start) / 1e6) / this.keys.length;
        if(show) {
          console.info(`Hazelcast read for size ${dataSize} bytes: ${averageTime.toFixed(6)} ms per operation`);
          fs.appendFile(this.resultFilename, 'Hazelcast, READ, ' + this.numOfInstances + ', ' + dataSize + ', ' + averageTime.toFixed(6) + '\n', (err) => {
            if(err) console.error(err);
          });
        }
  
        resolve()
      })
    })
  }
}

class SilentLogger {
  trace() {}
  debug() {}
  info() {}
  warn() {}
  error() {}
  log() {}
}

module.exports = HazelcastClient;

File: python_clients/hazelcast_client.py
from hazelcast import HazelcastClient as hazelcastClient
import random
import time

class HazelcastClient:
  def __init__(self, numOfInstances = 1, resultFilename = 'res-hazelcast-py.txt', totalOperations = 10000, totalProcesses = 1):
    self.numOfInstances = numOfInstances
    self.config = self.generateConfig(self.numOfInstances)
    self.resultFilename = resultFilename
    self.totalOperations = totalOperations
    self.totalProcesses = totalProcesses
    self.keys = []
    self.client = None
    self.map = None

  def generateConfig(self, numOfInstances):
    config = []

    # for i in range(numOfInstances):
    config.append(f'192.168.1.7:5700')
    if numOfInstances > 1: config.append(f'192.168.1.110:5701')
    if numOfInstances > 2: config.append(f'192.168.1.9:5702')

    print(config)
    
    return config

  def connect(self):
    self.client = hazelcastClient(
      cluster_members = self.config
    )

    self.map = self.client.get_map('benchmark-map')

    # print("Connected to Hazelcast")

  # def connect(self, instanceNumber):
  #   self.client = hazelcastClient(
  #     cluster_members = [f'127.0.0.1:570{instanceNumber}']
  #   )

  #   self.map = self.client.get_map('benchmark-map')

  def disconnect(self):
    self.client.shutdown()
    # print("Disconnected from Hazelcast")

  def benchmarkWrite(self, dataSize, show=False):
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.map.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Hazelcast WRITE SINGLE {dataSize} {averageTime:.8f}')
    
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Hazelcast WRITE SINGLE {dataSize} {averageTime:.8f}\n')
    resultFile.close()

  def benchmarkClusterWrite(self, processNum, totalProcesses):
    totalTimeTaken = 0
    data = 'a' * (1024 // totalProcesses)

    for i in range(self.totalOperations):
      key = f'key-{i}-{processNum}'
      start = time.time()
      self.map.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations

    print(f'Hazelcast WRITE CLUSTER {averageTime:.8f}')
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Hazelcast WRITE CLUSTER {averageTime:.8f}\n')
    resultFile.close()

  def benchmarkClusterRead(self, processNum, totalProcesses):
    totalTimeTaken = 0

    for i in range(self.totalOperations):
      key = f'key-{i}-{processNum}'
      start = time.time()
      self.map.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations

    print(f'Hazelcast READ CLUSTER {averageTime:.8f}')
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Hazelcast READ CLUSTER {averageTime:.8f}\n')
    resultFile.close()

  def benchmarkWriteMultiprocess(self, dataSize, show=False):
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.map.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Hazelcast WRITE MULTI {dataSize} {averageTime:.8f}')
    
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Hazelcast WRITE MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()


  def benchmarkWriteMultiprocessOld(self, dataSize, offset, show=False, processNum=1):
    self.connect()
    data = 'a' * dataSize

    totalTimeTaken = 0
    for i in range(offset, offset + self.totalOperations//self.totalProcesses):
      key = f'key-{dataSize}-{i}'
      self.keys.append(key)

      start = time.time()
      self.map.set(key, data)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Hazelcast WRITE MULTI {dataSize} {averageTime:.8f}')
    
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Hazelcast WRITE MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()

    self.disconnect()

  def benchmarkRead(self, dataSize, show=False):
    random.shuffle(self.keys)

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = self.keys[i]
      start = time.time()
      self.map.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Hazelcast READ SINGLE {dataSize} {averageTime:.8f}')
      
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Hazelcast READ SINGLE {dataSize} {averageTime:.8f}\n')
    resultFile.close()


  def benchmarkReadMultiprocess(self, dataSize, show=False, processNum=1):
    self.connect()
    random.shuffle(self.keys)

    totalTimeTaken = 0
    for i in range(self.totalOperations):
      key = self.keys[i]
      start = time.time()
      self.map.get(key)
      end = time.time()

      totalTimeTaken += end - start

    averageTime = totalTimeTaken / self.totalOperations
    if show:
      print(f'Hazelcast READ MULTI {dataSize} {averageTime:.8f}')
    
    resultFile = open(self.resultFilename, 'a')
    resultFile.write(f'Hazelcast READ MULTI {dataSize} {averageTime:.8f}\n')
    resultFile.close()

    self.disconnect()

  def cleanup(self):
    self.map.clear()
      